enhancements/ingress/custom-route-configuration.md

```diff
@@ -0,0 +1,242 @@
+    RelatedObjects []ObjectReference
+    
+    
+
+    // This API does not include a mechanism to distribute trust, since the ability to write this resource would then 
+    // allow interception.  Instead, if we need such a mechanism, we can talk about creating a way to allow narrowly scoped
+    // updates to a configmap containing ca-bundle.crt for each CustomizeableRoute.
+    // CurrentCABundle []byte
+}
+```
+
+
+### User Stories
+
+#### Cluster-admin wants to customize a route
+To use this, a cluster-admin would
+1.  Either read docs to find the namespace,name tuples or look at an existing cluster and read ingresses.config.openshift.io.
+2.  Create the serving cert/key pair secret in ns/openshift-config
+3.  Create an entry in `ingresses.config.openshift.io.spec.customizeableRouteSpec` for their customization
+4.  Wait to see the corresponding change in status.
+
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Control loop to manage secret read permissions in openshift-config
+List/watch for individual names has been possible for several releases.
+A control loop in either the ingress operator or the cluster config operator will watch for ingress.spec.customizeableRouteSpec.servingCertKeyPairSecret
+and will create a role/rolebinding for the corresponding ingress.status.customizeableRouteStatus.consumingUsers.
+This will allow an operator (or other binary with that user) to get/list/watch on the particular secret.
+
+This means that the power to mutate ingresses.config.openshift.io and ingresses.config.openshfit.io/status
+will imply the power to read secrets in openshift-config.
+If we decide we do not wish to do this, then it will incumbent upon the cluster-admin to create the role and rolebinding.
+
+#### library-go usage
+Because this will be leveraged by the authentication operator (and possibly the console operator), we can provide a config
+observer that reads the ingresses.spec, handles the specified names, sets up the specific list/watch, and takes care of
+copying the secret so it can be mounted.
+
+It would also be possible (though I'm less sure someone would use it), to provide a route injection option for serving cert/key
+pairs.
+
+Library-go is also an ideal spot to provide functionality to provide request route hosts.
+
+### Risks and Mitigations
+
+Some components have attempted to resolve similar issues in one off ways.
+This has produced a disjointed experience across the stack and challenging documentation to follow.
+Since these other ways exist, a migration to a unified mechanism may be challenging for those components.
+The helpers suggested above can ease that transition, but it still exists.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Which components should migrate?
+2. What timeline should that migration happen?
+3. Which components can start from this unified API?
+4. Do we allow OLM operators to participate?
```

Comment:
I have doubts about having both auto-RBAC assign and allowing OLM in.

Reply:
but technically nothing stops them, just permissions.

Reply:
At this time, OLM [does not support shipping routes](https://github.com/operator-framework/operator-registry/blob/master/pkg/lib/bundle/supported_resources.go#L27-L44) in Operator Bundles, so the operator would need to implement support for this feature.

OLM could certainly add support mind you.

Reply:
What does this mean for security? If I can write Ingress status, I can let me give permissions to any openshift-config secret? @deads2k this needs a discussion why this is ok or how to mitigate the problem

Reply:
That point is called out in the enhancement [here](https://github.com/openshift/enhancements/pull/577/files#diff-5a2afb792e11918fa5b74f9d217fc641a3af897066968cd2449e373f25f964abR183-R185) but doesn't seem to have garnered much conversation.

>This means that the power to mutate ingresses.config.openshift.io and ingresses.config.openshfit.io/status
will imply the power to read secrets in openshift-config. If we decide we do not wish to do this, then it will incumbent upon the cluster-admin to create the role and rolebinding.

Reply:
Note that you need to write both ingresses.config.openshift.io and ingresses.config.openshift.io/status.